### PR TITLE
📈(project) Add optional sentry integration

### PR DESCRIPTION
**Pull request by Sifflex**
## Purpose

As Ralph will manipulate sensible data, we should carefully watch jobs execution. Sentry will be a great help to detect failures.

## Proposal

We added a sentry integration which keeps track of every exception that occurs in the application

### CODE REVIEW

File: src/ralph/defaults.py [hunk 1]
```diff
@@ -37,3 +37,4 @@ class StorageBackends(Enum):
 ENVVAR_PREFIX = "RALPH"
 HISTORY_FILE = Path(environ.get("RALPH_HISTORY_FILE", APP_DIR / "history.json"))
 FS_STORAGE_DEFAULT_PATH = Path("./archives")
+SENTRY_DSN = None
```
Comment (jmaupetit) [inline src/ralph/defaults.py:40]: You should allow to override it using the environment.

File: src/ralph/cli.py [hunk 2]
```diff
@@ -32,6 +34,8 @@
 @click_log.simple_verbosity_option(get_root_logger())
 def cli():
     """Ralph is a stream-based tool to play with your logs"""
+    if SENTRY_DSN is not None:
+        sentry_sdk.init(SENTRY_DSN, traces_sample_rate=1.0)
```
Comment (jmaupetit) [inline src/ralph/cli.py:38]: I would recommend adding more configurations as:

- the `release`
- the `environment`

Reference documentation: https://docs.sentry.io/platforms/python/configuration/options/

**Review by jmaupetit — COMMENTED** [review]
Have you seen that sentry integrates with the logging module to add log messages as breadcrumbs on failure?

Reference: https://docs.sentry.io/platforms/python/guides/logging/

File: src/ralph/defaults.py (hunk 1 above)
Comment (Sifflex) [reply src/ralph/defaults.py:40]: Thanks, I forgot to add that

File: src/ralph/cli.py (hunk 2 above)
Comment (Sifflex) [reply src/ralph/cli.py:38]: Indeed, I added them

File: src/ralph/defaults.py [hunk 3]
```diff
@@ -101,3 +101,5 @@ def config(key, default_value):
 )
 HISTORY_FILE = Path(config("RALPH_HISTORY_FILE", APP_DIR / "history.json"))
 LOGGING_CONFIG = config("RALPH_LOGGING", DEFAULT_LOGGING_CONFIG)
+SENTRY_DSN = config("RALPH_SENTRY_DSN", None)
+ENVIRONMENT = config("RALPH_ENVIRONMENT", "DEVELOPMENT")
```
Comment (jmaupetit) [inline src/ralph/defaults.py:105]: Environment may be confusing (may refer to environment variables), maybe prefer "execution environment"?

```suggestion
EXECUTION_ENVIRONMENT = config("RALPH_EXECUTION_ENVIRONMENT", "development")
```

**Review by jmaupetit — COMMENTED** [review]
Great. Almost done.

File: src/ralph/__main__.py [hunk 4]
```diff
@@ -6,6 +6,19 @@
 This script is a POC for now, we aim to make ralph a nice CLI to use.
 """
 
-from . import cli
+import sentry_sdk
+
+from ralph.defaults import ENVIRONMENT, SENTRY_DSN
```
Comment (jmaupetit) [inline src/ralph/__main__.py:11]:
```suggestion
from ralph.defaults import EXECUTION_ENVIRONMENT, SENTRY_DSN
```

File: src/ralph/__main__.py [hunk 5]
```diff
@@ -6,6 +6,19 @@
...
+
+from . import __version__, cli
+
+if SENTRY_DSN is not None:
+    sentry_sdk.init(
+        dsn=SENTRY_DSN,
+        traces_sample_rate=1.0,
+        release=__version__,
+        environment=ENVIRONMENT,
```
Comment (jmaupetit) [inline src/ralph/__main__.py:20]:
```suggestion
        environment=EXECUTION_ENVIRONMENT,
```

**Review by jmaupetit — APPROVED** [review]

**Pull request merged**